src/track/track_gps_l2cm.c

```diff
@@ -0,0 +1,638 @@
+}
+
+/** Check if a satellite is already in track
+ *
+ * \param sat Satellite ID
+ * \retval false Satellite is not in track
+ * \retval true Satellite is in track
+ */
+static bool is_in_track(u16 sat)
+{
+  bool ret = false;
+  gps_l2cm_tracker_data_t* ptr;
+  u32 i;
+
+  for (i = 0;
+       i < sizeof(gps_l2cm_trackers) / sizeof(gps_l2cm_trackers[0]);
+       i++) {
+
+    if (!gps_l2cm_trackers[i].active) {
+      continue;
+    }
+    ptr = (gps_l2cm_tracker_data_t*)gps_l2cm_trackers[i].data;
+    if (NULL == ptr) {
+      continue;
+    }
+    if (ptr->sat == sat) {
+      ret = true;
+      break;
+    }
+  }
+  return ret;
+}
+
+/** Do L1C/A to L2 CM handover.
+ *
+ * The condition for the handover is the availability of bitsync on L1 C/A
+ *
+ * \param sample_count NAP sample count
+ * \param sat L1C/A Satellite ID
+ * \param nap_channel Associated NAP channel
+ * \param code_phase L1CA code phase [chips]
+ */
+void do_l1ca_to_l2cm_handover(u32 sample_count,
+                              u16 sat,
+                              u8 nap_channel,
+                              float code_phase)
+{
+  if (is_in_track(sat)) {
+    return; /* L2C signal from the SV is already in track */
+  }
+
+  /* First, get L2C capability for the SV from NDB */
+  u32 l2c_cpbl;
+  // TODO: uncomment this as soon as NDB gets available
+  // ndb_gps_l2cm_l2c_cap_read(&l2c_cpbl);
+
+  // TODO: remove this as soon as NDB gets available
+  // GPS PRNs with L2C capability:
+  // 01 03 05 06 07 08 09 10 12 15 17 24 25 26 27 29 30 31 32
+  // as per http://tinyurl.com/zj5q62h
+  l2c_cpbl = (u32)1 << (1 - 1);
+  l2c_cpbl |= (u32)1 << (3 - 1);
+  l2c_cpbl |= (u32)1 << (5 - 1);
+  l2c_cpbl |= (u32)1 << (6 - 1);
+  l2c_cpbl |= (u32)1 << (7 - 1);
+  l2c_cpbl |= (u32)1 << (8 - 1);
+  l2c_cpbl |= (u32)1 << (9 - 1);
+  l2c_cpbl |= (u32)1 << (10 - 1);
+  l2c_cpbl |= (u32)1 << (12 - 1);
+  l2c_cpbl |= (u32)1 << (15 - 1);
+  l2c_cpbl |= (u32)1 << (17 - 1);
+  l2c_cpbl |= (u32)1 << (24 - 1);
+  l2c_cpbl |= (u32)1 << (25 - 1);
+  l2c_cpbl |= (u32)1 << (26 - 1);
+  l2c_cpbl |= (u32)1 << (27 - 1);
+  l2c_cpbl |= (u32)1 << (29 - 1);
+  l2c_cpbl |= (u32)1 << (30 - 1);
+  l2c_cpbl |= (u32)1 << (31 - 1);
+  l2c_cpbl |= (u32)1 << (32 - 1);
+
+  /* compose SID: same SV, but code is L2 CM */
+  gnss_signal_t sid = {
+    .sat  = sat,
+    .code = CODE_GPS_L2CM
+  };
+
+  if (0 == (l2c_cpbl & ((u32)1 << (sat - 1)))) {
+    log_info_sid(sid, "SV does not support L2C signal");
+    return;
+  }
+
+  if ((code_phase < 0) ||
+      ((code_phase > 0.5) && (code_phase < (GPS_L1CA_CHIPS_NUM - 0.5)))) {
+    log_warn_sid(sid, "Unexpected L1C/A to L2C handover code phase: %f",
+                 code_phase);
+    return;
+  }
+
+  if (code_phase > (GPS_L1CA_CHIPS_NUM - 0.5)) {
+    code_phase = 2 * GPS_L2CM_CHIPS_NUM - (GPS_L1CA_CHIPS_NUM - code_phase);
+  }
+
+  /* recalculate doppler freq for L2 from L1*/
+  double carrier_freq = tracking_channel_carrier_freq_get(nap_channel) *
```

Comment:
I will replace elevation with TRACKING_ELEVATION_UNKNOWN assuming that manage_tracking_startup() will take care of it.
I am a bit confused about your comment regarding the doppler and cn0. If we know the doppler on L1C/A, it cannot be used as is for L2C tracker initialization because L2C has a different carrier frequency. So L1C/A doppler value must be scaled and the scaling factor is GPS_L2_HZ / GPS_L1_HZ. That is what the code is doing here. CN0 init value is taken to be the same as CN0 of L1C/A as the best estimation of CN0 at this stage.
So could you please elaborate?